### ISSUE TITLE
ChefSpec hangs..

### ISSUE THREAD

**Issue by rdean716** [sethvargo/chefspec#784]
We've been experiencing an issue during the execution of the unit tests where the process hangs. I have to break out of the process to regain control. This happens on both Windows 10 with Chefdk 1.03, Windows 2012 r2 with Chefdk 1.03, it was also happening on both platform with previous versions of Chefdk installed. On my build server, the 20120 R2 system, I've noticed that there's a ruby process running 50% CPU while the test is hung. The test hangs at different times in the test run, meaning it is not hanging on the same test or resource each time. Through process of elimination, I've removed each single test, and the process still hangs at different times,

I have no idea how to troubleshoot this further. Any thoughts?

**Comment by coderjoe** [comment]
I'm experiencing the same problem. Intermittent hangs while running chefspec on the latest stable Chefdk 1.1.16

**Comment by eehaight** [comment]
I am also experiencing this, along with extreme memory usage at the time of hanging. I have not been able to find a consistent reproducible test case

**Comment by keithbaker** [comment]
I'm also experiencing unit tests hanging in ChefDK 1.1.16 on windows 10.

**Comment by nmscholl** [comment]
I am experiencing this issue also. Intermittent hangs while running chefspec on the latest stable Chefdk 1.1.16, Windows 8.
Chef Development Kit Version: 1.1.16
chef-client version: 12.17.44
delivery version: master (83358fb62c0f711c70ad5a81030a6cae4017f103)
berks version: 5.2.0
kitchen version: 1.14.2

**Comment by rdean716** [comment]
I worked around this issue by using the Solorunner rather than the ServerRunner. The result is no hangs and significantly improved test execution time.

**Comment by coderjoe** [comment]
@rdean716 do you happen to have a reproducible test case that could be used to debug, test, and potentially fix the serverspec problem?

**Comment by rdean716** [comment]
@coderjoe I have one that I can reproduce easily on-site.  Not sure that I can post it without sanitizing it some first.

**Comment by coderjoe** [comment]
@rdean716 yeah I ended up doing something similar. Recommending a switch to SoloRunner to our developers. I imagine a reproducible test case would be welcome if your sanitization is successful.

**Comment by afiune** [comment]
Hello Chef Friends! 👋

Thank you for opening this issue, I am trying to replicate this behavior locally so any example or reproducible steps will be appreciated. I am currently working with one of the cookbooks that @rdean716 shared with us (Thank you!) and I can give a couple of ideas to work this out for the time being:

* One workaround is what @rdean716 recommended, by using `SoloRunner` you are not mocking the Chef Server so this should be much faster than the `ServerRunner`.
* If you are unable to switch to `SoloRunner` because you need to emulate the Chef Server (for any reason), then your option is to **cache** the `chef_run` execution so you do not create a completely new mocked ChefServer on every test, this should improve the execution of your test considerably. 

#### Here an example of the second suggestion
If you have the following test file:
```ruby
describe 'my awesome test suite' do
  let(:chef_run) do
    runner = ChefSpec::ServerRunner.new(platform: 'windows', version: '2012R2')
    runner.converge('test_cookbook::recipe')
  end

  it 'test1' do
    expect(chef_run).to create_directory('C:/my_app')
  end

  it 'test2' do
    expect(chef_run).to do_something('cool')
  end

  <... lets imagine you have 20 extra it statements similar as the above one ....>

end
```
You are going to create 20 `ChefZero` processes which needs to be initialized by registering the mocked node, uploading the cookbooks your cookbook depends on and a couple of extra tasks.

My recommendation would be to **cache** the `chef_run` as much as you can, I know there would be some times that you need a fresh mocked server but if you don't need that, then just cached! 😄 Just substitute the `let` statement for `cached` as follows:
```ruby
  cached(:chef_run) do
    runner = ChefSpec::ServerRunner.new(platform: 'windows', version: '2012R2')
    runner.converge('test_cookbook::recipe')
  end
```

This is not the final solution but it will help you a lot until I get to the bottom of the problem that causes the process to hang. I'll keep you posted! 👍

**Comment by rdean716** [comment]
@afiune, thanks for the feedback on this.  Adding the ```cached``` statement does significantly improve both the speed and reliability of the test execution.  I've noticed that the tests will hang still but it occurs much less frequently.

**Comment by tas50** [comment]
This should be fixed on master now with #809. Expect a new release tomorrow with that change.

**Comment by roberto-mardeni** [comment]
The change to SoloRunner definitely worked for me in the same setup, thank you @rdean716
